src/provider/meta.ts

```diff
@@ -0,0 +1,63 @@
+import type { ExternalProvider, JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+import WalletConnectProvider from '@walletconnect/ethereum-provider'
+
+function isWeb3Provider(provider: JsonRpcProvider): provider is Web3Provider {
+  return 'provider' in provider
+}
+
+function isWalletConnectProvider(provider: ExternalProvider): provider is WalletConnectProvider {
+  return (provider as WalletConnectProvider).isWalletConnect
+}
+
+export enum WalletType {
+  WALLET_CONNECT = 'WalletConnect',
+  INJECTED = 'Injected',
+  UNKNOWN = 'Unknown',
+}
+
+export interface WalletMeta extends Partial<NonNullable<WalletConnectProvider['connector']['peerMeta']>> {
+  type: WalletType
+}
+
+function getWalletConnectMeta(provider: WalletConnectProvider): WalletMeta {
+  return {
+    type: WalletType.WALLET_CONNECT,
+    ...provider.connector.peerMeta,
+  }
+}
+
+function getInjectedMeta(provider: ExternalProvider & Record<string, unknown>): WalletMeta {
+  const properties = Object.getOwnPropertyNames(provider)
+  const names =
+    properties
+      .filter((name) => name.match(/^is.*$/) && (provider as Record<string, unknown>)[name] === true)
+      .map((name) => name.slice(2)) ?? []
+
+  // Add qrUrl to identify QR-initiated connections (eg for Coinbase Wallet via mobile QR).
+  if (properties.includes('qrUrl') && provider['qrUrl']) {
+    names.push('qrUrl')
+  }
```

Comment:
Is this how WalletConnect does that? Any reference available or just to preface this function with the fact we're doing WalletConnect-ish compatibility here?

Reply:
This is done by coinbase. I will make that more explicit in the comment.